### PR TITLE
Fix Download and Mocking LLMLocalSession

### PR DESCRIPTION
**Pull request by LeonNissen**
# Fix Download and Mocking LLMLocalSession

## :gear: Release Notes 
- Resolved a download bug that caused the app to crash on the simulator after completing the download. The model is no longer initialized after the download finishes.
- Fixed the mock environment issue: the LLMLocalSession now generates a default message when used on the simulator.

## :pencil: Code of Conduct & Contributing Guidelines 

By submitting creating this pull request, you agree to follow our [Code of Conduct](https://github.com/StanfordSpezi/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordSpezi/.github/blob/main/CONTRIBUTING.md):
- [x] I agree to follow the [Code of Conduct](https://github.com/StanfordSpezi/.github/blob/main/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/StanfordSpezi/.github/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: Sources/SpeziLLMLocal/LLMLocalSession+Setup.swift
Comment (PSchmiedmayer) [inline Sources/SpeziLLMLocal/LLMLocalSession+Setup.swift:77]:
```suggestion
        try? await Task.sleep(for: .seconds(1))
```

File: Sources/SpeziLLMLocal/Mock/LLMLocalMockSession.swift
Comment (PSchmiedmayer) [inline Sources/SpeziLLMLocal/Mock/LLMLocalMockSession.swift:18]: Technically that would be considered a breaking change and we would need to tag a 0.10.0 for that.

**Review by PSchmiedmayer — APPROVED** [review]
Thank you for the improvements @LeonNissen!

File: Sources/SpeziLLMLocal/Mock/LLMLocalMockSession.swift
Comment (philippzagar) [reply Sources/SpeziLLMLocal/Mock/LLMLocalMockSession.swift:18]: I agree, we could wait until #64 is merged and then tag a 0.10.0

File: Sources/SpeziLLMLocal/LLMLocalSession+Generate.swift
Comment (philippzagar) [inline Sources/SpeziLLMLocal/LLMLocalSession+Generate.swift:158]: Maybe we could do all of this in a loop to prevent repetition? I remember writing that code a long time ago and not liking it back then, we could clean it up a bit.

File: Sources/SpeziLLMLocalDownload/LLMLocalDownloadManager.swift
Comment (philippzagar) [inline Sources/SpeziLLMLocalDownload/LLMLocalDownloadManager.swift:82]: Maybe we should make this function `async` at some point? Then we could use the SpeziViews `AsyncButton` to trigger the start of the download and prevent all the unnecessary `Task { @MainActor in ... }` scheduling.

**Review by philippzagar — APPROVED** [review]
Looks great! Just had a couple of ideas for follow up improvements! 🚀

File: Sources/SpeziLLMLocal/LLMLocalSession+Generate.swift
Comment (LeonNissen) [reply Sources/SpeziLLMLocal/LLMLocalSession+Generate.swift]: Yeah, sounds good, done ✅

File: Sources/SpeziLLMLocalDownload/LLMLocalDownloadManager.swift
Comment (LeonNissen) [reply Sources/SpeziLLMLocalDownload/LLMLocalDownloadManager.swift:82]: But don't we still need `Task { @MainActor in ... }` whenever we want to change `state` since it is `@MainActor`. So the only task that we could get rid of would be the download task, but than we might not be able to cancel the task.

Comment (philippzagar) [reply Sources/SpeziLLMLocalDownload/LLMLocalDownloadManager.swift:82]: Well, if the function is `async`, you can do `await MainActor.run {}`, meaning you actually await the execution of the closure, not just schedule one via `Task { @MainActor in ... }` (no proper cancellation, no guarantee of execution etc.)
We don't have to do that, just thought that this might be a nice addition.
Other option is to make the entire `startDownload()` function `@MainActor` (no `async` needed), but I'm not sure if that's desirable.
I'm also totally fine with leaving the code as is!

Comment (jdisho) [reply Sources/SpeziLLMLocalDownload/LLMLocalDownloadManager.swift:82]: @philippzagar - Interesting point here. I was imagining something like this, and it's kind of neat though 😄

```swift
func startDownload() async {
    if modelExists {
        await MainActor.run { state = .downloaded }
        return
    }
        
    cancelDownload()

    downloadTask = Task(priority: .userInitiated) {
        do {
            try await downloadModel()
            await MainActor.run { state = .downloaded }
        } catch {
            await MainActor.run {
                state = .error(
                    AnyLocalizedError(
                        error: error,
                        defaultErrorDescription: LocalizedStringResource(
                                "LLM_DOWNLOAD_FAILED_ERROR",
                                bundle: .atURL(from: .module)
                            )
                        )
                    )
                }
            }
        }

        await downloadTask?.value
    }
    
    @MainActor
    private func downloadModel() async throws {
        let repo = Hub.Repo(id: model.hubID)
        let modelFiles = ["*.safetensors", "config.json"]
        
        try await HubApi.shared.snapshot(from: repo, matching: modelFiles) { progress in
            self.state = .downloading(progress: progress)
        }
    }
```

Comment (jdisho) [reply Sources/SpeziLLMLocalDownload/LLMLocalDownloadManager.swift:82]: On `cancelDownload` we should update the state to `idle`, right?

File: Sources/SpeziLLM/Mock/LLMMockSession.swift
Comment (jdisho) [inline Sources/SpeziLLM/Mock/LLMMockSession.swift:63]: nit: I usually try to void double negations since they make the code harder to read. Instead we can use `if`.

**Review by jdisho — COMMENTED** [review]
Hey 👋 - this is my first time reviewing a PR here. I just had some ideas and suggestions, but nothing blocking.  

Really cool stuff in here ⚡️

**Comment by philippzagar** [comment]
> Hey 👋 - this is my first time reviewing a PR here. I just had some ideas and suggestions, but nothing blocking.
> 
> Really cool stuff in here ⚡️

Hey @jdisho, thanks for adding another review to this PR, great to have you on board! Feel free to let me know if I can help with anything!

File: Sources/SpeziLLM/Mock/LLMMockSession.swift
Comment (philippzagar) [reply Sources/SpeziLLM/Mock/LLMMockSession.swift:63]: We used this pattern throughout the code base, might make sense to simplify it. Definitely no must tho @LeonNissen

File: Sources/SpeziLLMLocalDownload/LLMLocalDownloadManager.swift
Comment (LeonNissen) [reply Sources/SpeziLLMLocalDownload/LLMLocalDownloadManager.swift:53]: Good catch, thanks! 😊

File: Sources/SpeziLLM/Mock/LLMMockSession.swift
Comment (LeonNissen) [reply Sources/SpeziLLM/Mock/LLMMockSession.swift:63]: Yeah exactly @philippzagar, but I see your point @jdisho. Maybe a todo for later?

File: Sources/SpeziLLMLocalDownload/LLMLocalDownloadManager.swift
Comment (LeonNissen) [reply Sources/SpeziLLMLocalDownload/LLMLocalDownloadManager.swift:82]: Thats an awesome solution thanks for providing this example! Its implemented now! 🚀

**Comment by LeonNissen** [comment]
Hey @jdisho, many thanks for the review! I've made the changes and fixed the typo. 🚀

**Comment by philippzagar** [comment]
@LeonNissen Let me know as soon as this PR is ready to merge, would be important to get this fix for the local platform in quickly!

**Pull request merged**
